### DISCUSSION TITLE
Modular VOLTTRON

### DISCUSSION THREAD

**Discussion by craigpnnl** [VOLTTRON/volttron#2824] [Ideas]
As I prepared for backporting the openadr-ven agent I came across dependency mismatch within the volttron environment.  VOLTTRON 8.x is currently pinned at cryptography==2.3 whilst the openadr-venv agent requires version 3.4.7.  Another issue is that I have a server running pythnon3.6 however the openadr-venv requires python3.8.  So I thought I would think about how we could solve these issues.  

1. Upgrade VOLTTRON
    - Seems obvious that we can work with a different version of cryptography, however right now VOLTTRON doesn't support this due to some certificate issues that are outside the scope of this post.
    - Upgrade VOLTTRON to use the new version of python.  Sure we can do this, but that means we have to install a new version of python and potentially install all the agents that are currently running.
2. Wrap a new environment around an installed agent
    - vctl install would then create a new python environment (symlinks to python to use) and then install the python agent in this new environment.  This is exactly what poetry does.  (https://python-poetry.org/).  There would be at least an additional file in the agent's directory to have the a reference to the venv.
    - vctl start would be able to use the correct python (even though it may/may not be the same python as the server) based upon the new venv.

The 2  above is actually already doable, though not directly with the modular code, however you would have to manually activate the environment and run vctl from that environment.  What this proposal is for is to make that process more inclusive in vctl.

Please let me know thoughts/criticisms to this plan.

**Comment by jkempf42** [comment]
Hi Craig,

If upgrading the platform to Python 3.8 isn't too time consuming, I think that would probably be the best solution because ultimately you will probably want to do that anyway. Alternatively, Solution 2 looks like a good choice. 

Hope that helps.

              jak

**Reply by lazlop** [reply]
Agreed! Option one seems like the better long-term solution as we will want to upgrade these things anyway.

**Comment by craigpnnl** [comment]
In talking with a lot of people we as a development team and community think that our time would be better spent to having tooling that will allow upgrading to happen out of band of the agent process.  In addition, there is a solution that is already present, just would need a developer to do the things I am talking about.

**Reply by amorphic** [reply]
As a relative newcomer to Volttron but a long-term Python user I would definitely advocate for Volttron to be _less_ active in creating and maintaining Python environments rather than more.
